### ISSUE TITLE
I tried your project and it doesn't work: It has a failure, i think related to competing threads.

### ISSUE THREAD

**Issue by djangofan** [virenv/ExtentReportUsage#4]
I tried your project and it doesn't work: It has a failure, i think related to competing threads.    I was hopeful at first that you had solved what I wasn't able to solve  ( i pasted the stack trace on this link) .   Here is my implementation of what you did using Gradle  :
https://github.com/djangofan/ParallelUsage-ExtentReports

Over a year ago, I tried something like what you did, not exactly the same but similar, in the past and was able to get the "parallel by classes" to work but had problems when running "parallel by methods".
https://github.com/djangofan/extent-reporting-example/tree/master/src/test/java/qa/test/tests

I notice you were not thinking of it in those terms.  You can run either of these above projects and you will see what I mean.

**Comment by virenv** [comment]
Hi @djangofan , The solution here is very old and might be failing because of radical changes in ExtentReports. However, the idea and logic that is described there can be extended to Class and method level. Can you give me the stack trace?

**Comment by djangofan** [comment]
Hi Virender,
Yeah, I put the stack trace in the README.md file in the project:
https://github.com/djangofan/ParallelUsage-ExtentReports

  I am really interested in this solution... you interesting in working
together on it at all?   I don't always have time, but every once in a
while I can dedicate time to it.   Like I said, Extent Reports was designed
to report when TestNG is ran on a per-class basis (as you can see on the
TestNG example on their site).    When I tried to do it by-method,  I ran
into some problems.   Maybe you know how to solve it?   I probably could
solve it given enough time.

-Jon

**Comment by anshooarora** [comment]
> Like I said, Extent Reports was designed
to report when TestNG is ran on a per-class basis (as you can see on the
TestNG example on their site).

I disagree with this statement.  It was designed to run in parallel in general,  the strategy doesn't matter.   See this for example:  http://extentreports.relevantcodes.com/java/#parallel-methods  which I feel can be refined further by adding `@BeforeMethod` and ending in `@AfterMethod`.

**Comment by djangofan** [comment]
Yes, your right.  Sorry,  last time I looked at that page (
http://extentreports.relevantcodes.com/java/#parallel-methods) ,  they
didn't have the parallelmethods example on there.  I sent them a message
about it (maybe a year ago)  and maybe he updated the page.

Given that new information, yes, I am hoping it will work.    I will give
it a try soon to make sure.

And maybe implementing their example wont throw that exception I was seeing?

-Jon

**Comment by djangofan** [comment]
Hmm... after looking at the example provided on his page (
http://extentreports.relevantcodes.com/java/#parallel-classes) ,  in his
example for parallel classes, he uses  the TestNG @AfterMethod hook rather
than @AfterClass .    I am not convinced it will work  but after I try it,
then I will know for sure.

**Comment by djangofan** [comment]
I guess I am saying :
1.  Problem with Scenario with a class containing more than 1 method.
Using parallel-classes, the report could flush out before the additional
methods are executed because of the use of @AfterMethod ?
2.  I think the scenario should therefore be adjusted to use @AfterClass
TestNG hook.
3.  I personally would design the ParallelClasses code to not allow a class
with more than 1 test method.      If a user has more than one test method,
then they should use parallel methods.

Just some thoughts.
-Jon

**Comment by virenv** [comment]
@djangofan Sure, we can work together to come up with a proper implementation. However, I recently had my friend create a framework that can use ExtentReports in any parallel scenario. It could be class or method level. The concept of ExtentTest should enable us do parallel testing at any level and that seems to be the original idea of @anshooarora . 

Looking at the stack trace and point number 1 that you had mentioned, Calling .flush will never be a problem. The problem should be calling .close on the report. So we should ideally do a .close at suite level. ISuiteListener.afterSuite is the most reliable method that I have see.

Lets work on this and lets build up a solution for your problem. Can you you share the code or is it the same where you have put the stack trace?

**Comment by virenv** [comment]
@djangofan with the example you have mentioned in here: https://github.com/djangofan/extent-reporting-example/tree/master/src/test/java/qa/test/tests

Parallel testing at test method level cannot be done. Your direct call to extentTest and methods like setTestName, setTestDescription can never work in Parallel environment.

Let me come up with a small implementation that I had been working on for one of my projects, I will share it with you guys.

**Comment by djangofan** [comment]
I haven't looked at that project in a while, but I remember I had it
working with parallel-by-class .  It was just the parallel-by-method that
was problematic.  If your saying it also doesn't  .setName  while in
parallel-by-class, that is news to me.  I think I had it working but now
that I look at it, I was also using @AfterMethod in my parallel by class
example, which doesn't seem right in retrospect.   I will wait and see what
you propose.

**Comment by virenv** [comment]
@djangofan .setName and similar .Set* methods are not thread safe and they are modifying variables which again are not thread safe. That way if you run them in parallel at method level, those will give unexpected results. When you run the test at parallel in class level, these will work that is because all the methods in the class will be run on the same thread in a sequence, so at any given instance only one test method will access the .set* methods. Anyway coming back, what are the requirement set. Can you please summarise?

**Comment by djangofan** [comment]
I see what your saying but if you generate threads via a testing data
provider then I think the testing factory creates a class instance on a per
method basis.  What your saying might be true if you fork test threads via
surefire or grade or otherwise.   As a sdet,  every project I've done has
been parallel by class.  The only think I want to know is achievable is:
 reporting with parallel methods.  So I do this only for learning
experience.

**Comment by virenv** [comment]
All right so lets make a true parallel ExtentReports wrapper that can work agnostic of a testing framework. I will add a project today with starter code today. We will gather more requirements and design from there.

**Comment by virenv** [comment]
@anshooarora I hope you know about this video: https://www.youtube.com/watch?v=DzVfainzmYo at 2:58

**Comment by djangofan** [comment]
Yeah, I know about ExtentX. I think we should support that , but I want to
focus on solving the parallel threading thing first. Do you agree? I think
you might be suggesting that ExtentX would be a workaround for the parallel
problems we might encounter?

**Comment by virenv** [comment]
@djangofan  that was for @anshooarora . How ExtentReports is becoming a standard for reporting infrastructure.

Anyway for you this the project that I have created https://github.com/virenv/ParallelReporter. Take a look at it, it might look verbose and has opportunity to optimise, but it will give you an idea on how you can achieve any level of parallelism in ExtentReports.

**Comment by anshooarora** [comment]
Thanks Viren,  I wasn't aware of that.  I will have a look tonight,  thanks much for sharing!

Have you guys used the framework that @saikrishna321 created here:  https://github.com/saikrishna321/AppiumTestDistribution

If you see the [reporting package](https://github.com/saikrishna321/AppiumTestDistribution/tree/master/src/main/java/com/report/factory), I had worked with him to setup a few things here that you might find helpful.

**Comment by djangofan** [comment]
I will look at it more in depth later today (since I took the day off to
take my Java cert exam).  I gave you a pull request , for the time being,
that has changes only to  non-Java project files:
https://github.com/virenv/ParallelReporter/pull/2
